File: meilisearch-utils/update-document.py
#!/usr/bin/env python3

import hashlib
import keyring
import meilisearch
import os
import re
import sys


## Config 

domain = "http://localhost:7700"

index_name = 'grimoire'

admin_key = keyring.get_password(
    'alan--meilisearch--scratchpad--admin-key',
    'alan'
)


## Functions

def string_to_md5(string):
    return hashlib.md5(bytes(string, encoding="utf-8")).hexdigest()


def do_update():
    if (len(sys.argv) > 1):

        update = True

        file_path = sys.argv[1]
        file_basename = os.path.basename(file_path).split('.')[0]
        file_extension = os.path.splitext(file_path)[1].lower()

        if file_extension != ".org":
            update = False

        if len(file_basename) == 0:
            update = False

        if len(file_basename) == 0 or re.search(r"^\W", file_basename[0]):
            update = False

        if update == True:

            print(f"Updating: {file_path}")

            with open(file_path, 'r', encoding='utf-8', errors='ignore') as _in:

                client = meilisearch.Client(
                    domain,
                    admin_key
                )

                filename = file_path.split('/')[-1]
                id_string = string_to_md5(file_path)
                content = _in.read()

                payload = {
                    "id": id_string,
                    "filename": filename,
                    "content": content
                }

                response_upload = client.index(index_name).add_documents(payload)

        else:
            print(f"Skipping: {file_path}")

if __name__ == "__main__":
    do_update()

File: meilisearch-updaters/src/update-document.js
#!/usr/bin/env node

import fs from 'fs';
import child_process from 'child_process';
import { FileParser } from './lib/FileParser.js';
import { MeiliSearch } from 'meilisearch'

/////////////////////////////////////////
// TODO: Move this to a config file

const rootDir = '/Users/alan/Grimoire';

const indexName = 'grimoire';

const host = 'http://127.0.0.1:7700';

const securityCall = `security find-generic-password -w -a alan -s alan--meilisearch--scratchpad--admin-key`;

/////////////////////////////////////////


if (process.argv.length > 1) {
    const filePath = process.argv[2];
    const filePathParts = filePath.split("/");
    const fileName = filePathParts.pop();
    const fileNameParts = fileName.split('.');
    if (fileNameParts.length > 1) {
        const fileExtension = fileNameParts.pop();
        if (fileExtension === 'org') {
            if (fileNameParts[0].match(/\w/)) {
                console.log(`Updating: ${filePath}`)
                const contents = fs.readFileSync(filePath, 'utf8');
                const payload = [];
                payload[0] = FileParser({ fileName: fileName, contents: contents });
                const apiKey = child_process
                      .execSync(securityCall, { encoding: 'utf-8' })
                      .trim();
                const client = new MeiliSearch({
                    host: host,
                    apiKey: apiKey,
                });

                // delete the id to deal with
                // renames. TODO: Figure out if this
                // is a problem if the document id
                // doesn't yet. 
                client
                    .index(indexName)
                    .deleteDocument(payload[0].id)

                client
                    .index(indexName)
                    .addDocuments(payload)
                // .then((res) => console.log(res));
            }
        }
    }
}

